### ISSUE TITLE
How to stop the player without destroying?

### ISSUE THREAD

**Issue by zoli456** [Tomato6966/lavalink-client#15]
I am moving away your erela.js. but I cant a way to stop the player without destroying it.

**Comment by zoli456** [comment]
My quick fix was copying over the old stop() method and adding a try{}catch(){} around **player.queue.previous.unshift(track);** in **queueEnd** function and empty the **player.queue** .
```java
new_player.stop = async function () {
      const now = performance.now();
      await this.node.updatePlayer({
        guildId: this.guildId,
        playerOptions: { encodedTrack: null },
      });
      this.ping = Math.round((performance.now() - now) / 10) / 100;
      return true;
    };
```

**Comment by zoli456** [comment]
There is another thing. It would be good to have a build-in support for the sponsor blocker plugin (https://github.com/topi314/Sponsorblock-Plugin).
```java
    async setsponsorblocker(segments = ["sponsor", "selfpromo"]){
        await fetch(`http://${this.node.options.host}:${this.node.options.port}/${this.node.options.version}/sessions/${this.node.sessionId}/players/${this.guild}/sponsorblock/categories`, {
            method: 'PUT',
            body: JSON.stringify(segments),
            headers: {Authorization: this.node.options.password, 'Content-Type': 'application/json'}
        });
    }
```
And need 2 new events
```java
 else if (payload.type === "SegmentsLoaded") {
            this.manager.emit("SegmentsLoaded", this, payload);
        }
        else if (payload.type === "SegmentSkipped") {
            this.manager.emit("SegmentSkipped", this, payload);
        }
```

**Comment by Tomato6966** [comment]
> There is an another thing. It would be good to have build-in support for the sponsor blocker plugin (https://github.com/topi314/Sponsorblock-Plugin).
> 
> ```java
>     async setsponsorblocker(segments = ["sponsor", "selfpromo"]){
>         await fetch(`http://${this.node.options.host}:${this.node.options.port}/${this.node.options.version}/sessions/${this.node.sessionId}/players/${this.guild}/sponsorblock/categories`, {
>             method: 'PUT',
>             body: JSON.stringify(segments),
>             headers: {Authorization: this.node.options.password, 'Content-Type': 'application/json'}
>         });
>     }
> ```
> 
> And need 2 new events
> 
> ```java
>  else if (payload.type === "SegmentsLoaded") {
>             this.manager.emit("SegmentsLoaded", this, payload);
>         }
>         else if (payload.type === "SegmentSkipped") {
>             this.manager.emit("SegmentSkipped", this, payload);
>         }
> ```

Thanks for that, I will take a look

**Comment by zoli456** [comment]
I am struggling to fix a problem in my bot. If a track trigger the `trackError` event and queue hasn't got more tracks left it send the `queueEnd` event  twice. If it has remaning tracks it not continue playing the queue with `autoSkip` enabled .

**Comment by Tomato6966** [comment]
> This library really need a Stop() function.

Implemented with version **1.2.0**:
`player.stopPlaying()` -> it will clear queue, and execute "stop" function for lavalink, to stop playing without destroying the player.
Also added to skip the "throwError" Property, so you can do:

`player.skip(0, false)` or shortcut: `player.skip(false)` which still skips, but if no tracks in queue it will not throw an error, instead execute "stop" and don't destroy the player..

**Note**: If the player isn't playing something and you execute `player.skip(0, false)` it will try to play the next song, which ofc. isn't possible and that will throw an error (via events)

For the trackError double execution of queueEnd, i'll take a look later today and if i can reproduce it, i will throw out a patch for that.

**Comment by Tomato6966** [comment]
> I am struggling to fix a problem in my bot. If a track trigger the `trackError` event and queue hasn't got more tracks left it send the `queueEnd` event twice. If it has remaning tracks it not continue playing the queue with `autoSkip` enabled .
> 
> I use long [Google TTS ](https://github.com/DuncteBot/skybot-lavalink-plugin) requests to trigger the error.

I can't figure out that problem / replicate it.
Can you declare what Event fires first and what next etc. etc.
I also can't test it using Google TTS Plugin since i Don't have a google tts api key. You may text me on discord for that (chrissy8283)

**Comment by zoli456** [comment]
I used the [DuncteBot lavalink plugin](https://github.com/DuncteBot/skybot-lavalink-plugin#dunctebot-lavalink-plugin). It doesnt need api key. You only need to enable the source and set the language and send request with `speak`. If you send a few hundred character long text it will trigger the error.
I sent a friend request on Discord.

**Comment by Tomato6966** [comment]
Well what you said was google tts, and google tts needs api key, duncte bot doesn't use google tts :)
Anyways.
it's not officially supported for lavalink v4. but i'll see if i can reproduce it with it What version are you running on?

**Comment by zoli456** [comment]
I am using PRE_1.5.0_25.

**Comment by zoli456** [comment]
`setSponsorBlock()` cause `Content-Type 'application/octet-stream' is not supported` error.

**Comment by Tomato6966** [comment]
> Author

Should be fixed with version 1.2.2

**Comment by Tomato6966** [comment]
It should be fixed with v1.2.3 that the queueEnd get's executed twice on trackError / trackStuck.
it's because after trackError trackEnd get's executed with the reason loadFailed.

> I am using [PRE_1.5.0_25](https://github.com/DuncteBot/skybot-lavalink-plugin/releases)
